File: src/com/castlefrog/agl/domains/biniax/BiniaxState.java
package com.castlefrog.agl.domains.biniax;

import com.castlefrog.agl.State;

/**
 * Defines a Biniax state.
 * Location value representation:
 * -1: impassable
 *  0: empty
 *  x: single element 0 < x <= MAX_ELEMENTS
 *  xy: element pair 0 < x <= MAX_ELEMENTS and 0 < y <= MAX_ELEMENTS and x < y
 */
public final class BiniaxState implements State<BiniaxState> {
    private static final int WIDTH = 5;
    private static final int HEIGHT = 7;
    private static final int MAX_ELEMENTS = 9;
    private static final int BUFFER = 3;
    public static final int ELEMENT_LIMIT = MAX_ELEMENTS + 1;
    public static final int MAX_FREE_MOVES = 2;

    private final byte[][] locations_;

    private final byte freeMoves_;
    /** Total number of actions taken. */
    private final int nTurns_;

    /**
     * The initial state for biniax starts with the player at the bottom of
     * the grid and some random elements above the player.
     */
    public BiniaxState() {
        locations_ = new byte[BiniaxState.getWidth()][BiniaxState.getHeight()];
        for (int i = 0; i < BiniaxState.getHeight(); i++) {
            int emptyLocation = (int) (Math.random() * BiniaxState.getWidth());
            for (int j = 0; j < BiniaxState.getWidth(); j++) {
                if (j != emptyLocation && i < BiniaxState.getHeight() - BUFFER) {
                    locations_[j][i] = (byte) BiniaxSimulator.generateRandomElementPair(0);
                    if (i == BiniaxState.getHeight() - BUFFER - 1) {
                        locations_[j][i] = (byte) (locations_[j][i] % ELEMENT_LIMIT + ELEMENT_LIMIT);
                    }
                }
            }
        }
        locations_[BiniaxState.getWidth() / 2][BiniaxState.getHeight() - 1] = 1;
        freeMoves_ = MAX_FREE_MOVES;
        nTurns_ = 0;
    }

    public BiniaxState(byte[][] locations, byte freeMoves, int nTurns) {
        locations_ = locations;
        freeMoves_ = freeMoves;
        nTurns_ = nTurns;
    }

    public BiniaxState copy() {
        return this;
    }

    public byte[][] getLocations() {
        byte[][] locations = new byte[WIDTH][HEIGHT];
        for (int i = 0; i < WIDTH; i++) {
            System.arraycopy(locations_[i], 0, locations[i], 0, HEIGHT);
        }
        return locations;
    }

    public byte getLocation(int x, int y) {
        return locations_[x][y];
    }

    public byte getFreeMoves() {
        return freeMoves_;
    }

    public int getNTurns() {
        return nTurns_;
    }

    //public int getNElementTypes() {
    //    return 0;
    //}

    public static int getWidth() {
        return WIDTH;
    }

    public static int getHeight() {
        return HEIGHT;
    }

    public static int getMaxElements() {
        return MAX_ELEMENTS;
    }

    @Override
    public int hashCode() {
        int code = 7 + freeMoves_;
        for (byte[] locations : locations_) {
            for (byte location : locations) {
                code = 11 * code + location;
            }
        }
        code = 11 * code + nTurns_;
        return code;
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof BiniaxState)) {
            return false;
        }
        BiniaxState state = (BiniaxState) object;
        if (freeMoves_ != state.getFreeMoves()) {
            return false;
        }
        for (int i = 0; i < WIDTH; i += 1) {
            for (int j = 0; j < HEIGHT; j += 1) {
                if (locations_[i][j] != state.getLocation(i, j)) {
                    return false;
                }
            }
        }
        return nTurns_ == state.getNTurns();
    }

    @Override
    public String toString() {
        final String elements = " ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        StringBuilder output = new StringBuilder();
        output.append("(Turns: ").append(nTurns_).append(")\n");
        output.append("(Free Moves:").append(freeMoves_).append(")\n");
        for (int i = 0; i < WIDTH; i += 1) {
            output.append("----");
        }
        output.append("-\n");
        for (int j = 0; j < HEIGHT; j += 1) {
            output.append(":");
            for (int i = 0; i < WIDTH; i += 1) {
                if (i != 0) {
                    output.append(" ");
                }
                if (locations_[i][j] == 0) {
                    output.append("   ");
                } else if (locations_[i][j] == -1) {
                    output.append("<X>");
                } else if (locations_[i][j] > 0 && locations_[i][j] < (MAX_ELEMENTS + 1)) {
                    output.append("[");
                    output.append(elements.charAt(locations_[i][j]));
                    output.append("]");
                } else {
                    output.append(elements.charAt(locations_[i][j] / (MAX_ELEMENTS + 1)));
                    output.append("-");
                    output.append(elements.charAt(locations_[i][j] % (MAX_ELEMENTS + 1)));
                }
            }
            output.append(":\n");
        }
        output.append("-");
        for (int i = 0; i < WIDTH; i++) {
            output.append("----");
        }
        return output.toString();
    }
}

File: src/main/kotlin/com/castlefrog/agl/domains/biniax/BiniaxState.kt
package com.castlefrog.agl.domains.biniax

import com.castlefrog.agl.State
import java.util.Arrays

/**
 * Defines a Biniax state.
 * Location value representation:
 * 0: empty
 * x: single element 0 < x <= MAX_ELEMENTS
 * xy: element pair 0 < x <= MAX_ELEMENTS and 0 < y <= MAX_ELEMENTS and x < y
 */
data class BiniaxState(
        val locations: ByteArray = ByteArray(BiniaxSimulator.WIDTH * BiniaxSimulator.HEIGHT),
        val maxElements: Byte = 10,
        var freeMoves: Byte = 2,
        var nTurns: Int = 0) : State<BiniaxState> {

    init {
        if (locations.size != BiniaxSimulator.WIDTH * BiniaxSimulator.HEIGHT) {
            throw IllegalArgumentException("locations size must be = " + BiniaxSimulator.WIDTH * BiniaxSimulator.HEIGHT)
        }
    }

    override fun copy(): BiniaxState {
        val copyLocations = ByteArray(locations.size)
        System.arraycopy(locations, 0, copyLocations, 0, locations.size)
        return copy(locations = copyLocations, maxElements = maxElements, freeMoves = freeMoves, nTurns = nTurns)
    }

    override fun hashCode(): Int {
        return 11 * (17 * freeMoves + Arrays.hashCode(locations)) + nTurns
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false
        return other is BiniaxState
                && Arrays.equals(locations, other.locations)
                && freeMoves == other.freeMoves
                && nTurns == other.nTurns
    }

    override fun toString(): String {
        val output = StringBuilder()
        output.append("Turns: ").append(nTurns).append("\n")
        output.append("Free Moves: ").append(freeMoves.toInt()).append("\n")
        for (i in 0 until BiniaxSimulator.WIDTH) {
            output.append("----")
        }
        output.append("-\n")
        for (j in 0 until BiniaxSimulator.HEIGHT) {
            output.append(":")
            for (i in 0 until BiniaxSimulator.WIDTH) {
                val location = locations[i + j * BiniaxSimulator.WIDTH].toInt()
                if (i != 0) {
                    output.append(" ")
                }
                when (location) {
                    0 -> output.append("   ")
                    in 1..(maxElements - 1) -> {
                        output.append("[")
                        output.append((0x40 + location).toChar())
                        output.append("]")
                    }
                    else -> {
                        output.append((0x40 + location / (maxElements)).toChar())
                        output.append("-")
                        output.append((0x40 + location % (maxElements)).toChar())
                    }
                }
            }
            output.append(":\n")
        }
        output.append("-")
        for (i in 0 until BiniaxSimulator.WIDTH) {
            output.append("----")
        }
        return output.toString()
    }

}
